editor/src/components/text-editor/text-editor.tsx

```diff
@@ -127,6 +158,25 @@ export const TextEditor: React.FC<TextEditorProps> = ({ elementPath, text }: Tex
     [dispatch, elementPath, allElementProps],
   )
 
+  const onPaste = React.useCallback(
+    (e: React.ClipboardEvent) => {
+      if (myElement.current == null || myElement.current.textContent == null) {
+        return
+      }
+
+      const pastedText = e.clipboardData.getData('text')
+
+      const before = myElement.current.textContent.slice(0, caretPosition)
+      const after = myElement.current.textContent.slice(caretPosition)
+      myElement.current.textContent = before + pastedText + after
+
+      const newCaretPosition = caretPosition + pastedText.length
+      moveCaret(myElement.current, newCaretPosition)
+      setCaretPosition(newCaretPosition)
+    },
+    [caretPosition],
+  )
```

Comment:
I'm not sure about this, but isn't that easier to force the browser to actually insert the pasted text with `document.execCommand('insertText', pastedText)` , just like in the tests?
And then we would need this whole complexity and manual cursor management.
BTW do we know why paste doesn't work out of the box? It should, shouldn't it?

Reply:
> isn't that easier to force the browser to actually insert the pasted text

sure, but this felt simple enough to justify not using a deprecated API

> BTW do we know why paste doesn't work out of the box?
no idea ¯\_(ツ)_/¯

Reply:
To be honest I would try to investigate first why it doesn't work (maybe we have a `preventDefault` somewhere?), because this solution feels more fragile than the native paste handling of the browser.

Reply:
incredibly enough, it all boiled down to just doing a `stopPropagation()` on the onPaste event o.O

https://github.com/concrete-utopia/utopia/pull/3017/commits/7e9c54c43aa66d118e590554c3dddbecfe991420

I'm now working on the tests which are not working anymore, but the functionality is there :)

Reply:
> BTW do we know why paste doesn't work out of the box?

I'm pretty sure it's because Utopia grabs top-level paste events so you can eg paste something onto the canvas even if eg the navigator is focused.